Kernel/CMakeLists.txt

```diff
@@ -559,6 +559,15 @@ endif()
 
 if ("${SERENITY_ARCH}" STREQUAL "aarch64")
     add_compile_options(-fno-threadsafe-statics)
+
+    # This is needed on aarch64, since the compiler will otherwise assume that
+    # unaligned memory accesses are handled by the kernel/hardware, however
+    # we enforce everything to be properly aligned. Thus we need this option
+    # to make sure that the compiler actually aligns all the data properly.
```

Comment:
FWIW, this sounds vaguely fishy to me. If we make sure everything is aligned, then why does the compiler have to do anything?

This comment isn't based in reality in any form really – I'm not sure what this flag does. But I know a few similar flags, and what they often mean is "assume user code is not aligned, but the system requires it, so insert fixups to make certain things are aligned".

What I think we want is to make our user code all aligned, so that even though the system doesn't handle unaligned accesses, we don't have any unaligned accesses since all our code and data is already aligned correctly.

(But again, this comment is based on feeling and experience with similar flags, not with this particular flag. So it's possible I'm off here.)

Reply:
Yes, I agree that we should make sure that the code itself actually make sure everything is properly aligned. This is where the misalignment is happening:

```
   0x00000000000939d8 <+232>:   b.ne    0x939c8 <Kernel::init()+216>  // b.any
   0x00000000000939dc <+236>:   bl      0xb7ae0 <_ZN6Kernel24load_kernel_symbol_tableEv>
   0x00000000000939e0 <+240>:   add     x23, sp, #0x70
   0x00000000000939e4 <+244>:   mov     x0, x23
   0x00000000000939e8 <+248>:   bl      0x92940 <_ZN6Kernel7Mailbox13MessageHeaderC2Ev>
   0x00000000000939ec <+252>:   add     x0, sp, #0x78
   0x00000000000939f0 <+256>:   mov     w2, #0x4                        // #4
   0x00000000000939f4 <+260>:   mov     w1, #0x1                        // #1
   0x00000000000939f8 <+264>:   bl      0x92920 <_ZN6Kernel7Mailbox7MessageC2Ejj>
   0x00000000000939fc <+268>:   stur    xzr, [sp, #132] <--------------------------------------------------------------------------
   0x0000000000093a00 <+272>:   mov     w19, #0xffffffff                // #-1
   0x0000000000093a04 <+276>:   bl      0x92970 <_ZN6Kernel7Mailbox3theEv>
   0x0000000000093a08 <+280>:   add     x24, sp, #0x98
   0x0000000000093a0c <+284>:   mov     x1, x23
   0x0000000000093a10 <+288>:   mov     w2, #0x20                       // #32
   0x0000000000093a14 <+292>:   bl      0x92980 <_ZNK6Kernel7Mailbox10send_queueEPvj>
```

Which is in this function:
https://github.com/SerenityOS/serenity/blob/857a767ab49b6813707b81d93fab81d524cca574/Kernel/Arch/aarch64/init.cpp#L138-L151

It's pretty janky, because the `MessageHeader` and `MessageTail` are just normal classes, no packed structs or anything. So a think something is going wrong here. For now UBSAN is enabled by default, which will get rid of the exception, but this code should still be fixed sometime.

Reply:
That's a lot of jank. If you need alignment, how about `max(sizeof(Mailbox::MessageHeader), max(sizeof(Mailbox::MessageTail), QueryFirmwareVersionMboxMessage))`, which is what the compiler should normally do? Or why is the align here in the first place, is that not done automatically?

Reply:
I wonder if we had this problem with the LHS of https://github.com/SerenityOS/serenity/commit/3cc5752a69eebc1a92d447b3bf254faaa5eafb7b#diff-25f1ccd9541ff907401a9a8b70840177a86b0a670509641f7ea7cf28c4feef51L82 …

Reply:
@nico Guess what, we don't have the problem when using the old code. So that is definitely the issue!